File: roa.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <ctype.h>

// Attention aux indiens  little indian vs big indian !!!
// ce code ne marche que sur du x86 (octets inversés) et affiliés
// et considère les tailles des types suivants (en octets)
// sizeof(int) = 4
// sizeof(long) = 8
// sizeof(float) = 4
// sizeof(double) = 8
//
// La lecture du fichier se fera Element par element, en effet
// il y a des incohérences entre les tailles des structures
// et le nombre d'element dans le fichier
// et aussi des bugs dans le fichier, par exemple 
// le bloc Stats est marqué dans le fichier comme ayant une longueur 20 octets
// alors qu'il en fait réellement 24 !!
//

int verbose = 0;
int opt_samples = 10;

#ifndef BYTE
#define BYTE 	int_least8_t
#endif

#ifndef WORD
#define WORD	int_least16_t
#endif

#ifndef DWORD
#define DWORD	int_least32_t
#endif

// from dat file format.doc : .  I believe TDateTime is a 4 byte double (in fact a float)
#ifndef TDateTime
#define TDateTime float
#endif
// Apres analyse, il s'agit plutôt d'un double de 8 octets

//Snapshot header
#pragma pack(push,1)
typedef struct
{
  BYTE VersionNum;
  BYTE FILE_SYNC[3];
}SNAPSHOT_FILE_HEADER;
#pragma pack(pop)

//Snapshot header
#pragma pack(push,1)
typedef struct
{
  WORD     SectionID;
  BYTE     VersionNum;
  BYTE     SyncByte;
  DWORD SectionLen;
}SNAPSHOT_HEADER;
#pragma pack(pop)

// program d'index des fichiers -----
#define MAX_SAMPLES  1024
#define DATA_FILE_VERSION  1

#define    SECTION_SYNC_BYTE       0x40

#define    LOCAL_DATA_SECTION_ID   0x1111
#define    REMOTE_DATA_SECTION_ID  0x2222
#define    STATS_SECTION_ID        0x3333
#define    DESCR_SECTION_ID        0x4444

#define   FILE_SYNC_SEQ           0x414B53


#pragma pack(push,1)
typedef struct tsDatFile {
   SNAPSHOT_FILE_HEADER snapshotFileHead;
   SNAPSHOT_HEADER snapshotDataHead;
   float ROAspecta[MAX_SAMPLES];
   float RAMANspecta[MAX_SAMPLES];
   double LeftHand[2][2][2][MAX_SAMPLES];
   double RightHand[2][2][2][MAX_SAMPLES];
   float ROA_CC[MAX_SAMPLES];
   float ROAnoiseFloor[MAX_SAMPLES];
   SNAPSHOT_HEADER snapshotStatsHead;
   int dataFileVersion;
   int numberOfScans; 		// au depart long
   double elapsedTime;		// au depart float
   double TotalExposureTime;
   SNAPSHOT_HEADER snapshotDescrHead;
   char sampleName[100];
   char sampleDescription[2000];
} TDataFile;
#pragma pack(pop)

void print_header(SNAPSHOT_HEADER * pheader)
{
	char * label;

	if (verbose < 1)
		return;

	switch(pheader->SectionID){
		case LOCAL_DATA_SECTION_ID :
			label = "local-data-section";
			break;
		case REMOTE_DATA_SECTION_ID :
			label = "remote-data-section";
			break;
		case STATS_SECTION_ID :
			label = "stats-section";
			break;
		case DESCR_SECTION_ID :
			label = "description-section";
			break;
		default:
			label = "<unknown>";
	}
	printf("%s : 0x%04x : VersionNum : 0x%02x : SyncByte : 0x%02x : SectionLen :  %d 0x%04x 0%8o\n",
		       	label , pheader->SectionID,
			pheader->VersionNum,
			pheader->SyncByte,
			pheader->SectionLen, pheader->SectionLen, pheader->SectionLen);
#if 0
	printf("%s : VersionNum : 0x%02x\n", label , pheader->VersionNum);
	printf("%s : SyncByte : 0x%02x\n",   label, pheader->SyncByte);
	printf("%s : SectionLen : %d 0x%04x %8o\n", label,
			pheader->SectionLen, pheader->SectionLen, pheader->SectionLen);
#endif
}

void print_samples(TDataFile * pd, int nbsamples)
{
	int i;
	if (nbsamples >= MAX_SAMPLES){
		fprintf(stderr,"WARNING: too much samples : %d\n", nbsamples);
		nbsamples = 10;
	}
	for (i = 0; i < nbsamples; i++){
		if (i == 0){
			printf("%-15s ; %-15s ; %-15s ; %-15s\n", "roa_cc", "roanoiseFloor", "raman", "roa");
		}
		printf("%15f ; %15f ; %15.f ; %15f\n", 
			pd->ROA_CC[i], pd->ROAnoiseFloor[i],
			pd->RAMANspecta[i],
			pd->ROAspecta[i]);
	}
	printf("\n");
}

void LoadSnapShot(char * fname)
{
	FILE * fd;
	TDataFile d;
	size_t ret;
	SNAPSHOT_HEADER header;
	size_t position;
	size_t old_position;

	printf("fname: %s\n", fname);

	fd = fopen(fname, "rb");
	if (!fd){
		fprintf(stderr,"%s : %d : %s\n", fname, errno, strerror(errno));
		return;
	}
	// File Header
        ret = fread(&d.snapshotFileHead, sizeof(d.snapshotFileHead), 1, fd);
	if (verbose)
		printf("VersionNum: %d\n", d.snapshotFileHead.VersionNum);
	if (d.snapshotFileHead.VersionNum != DATA_FILE_VERSION ){
		fprintf(stderr, "WARNING: bad version number in file: '%s', bad format.\n", fname);
	}
	position = sizeof(d.snapshotFileHead);

	while( fread(&header, sizeof(header), 1, fd) == 1 ){ 
		
		position += sizeof(header);
		position += header.SectionLen; // position théorique du prochain header
		print_header(&header);
		switch(header.SectionID) {
			case LOCAL_DATA_SECTION_ID :
				memcpy(&d.snapshotDataHead, &header, sizeof(header));
				ret = fread(d.ROAspecta, sizeof(d.ROAspecta), 1, fd);
				ret = fread(d.RAMANspecta, sizeof(d.RAMANspecta), 1, fd);
				ret = fread(d.LeftHand, sizeof(d.LeftHand), 1, fd);
				ret = fread(d.RightHand, sizeof(d.RightHand), 1, fd);
				ret = fread(d.ROA_CC, sizeof(d.ROA_CC), 1, fd);
				ret = fread(d.ROAnoiseFloor, sizeof(d.ROAnoiseFloor), 1, fd);

				print_samples(&d, opt_samples);

				break;
			case STATS_SECTION_ID :
				memcpy(&d.snapshotStatsHead, &header, sizeof(header));
				ret = fread(&d.dataFileVersion, sizeof(d.dataFileVersion), 1, fd);
				ret = fread(&d.numberOfScans, sizeof(d.numberOfScans), 1, fd);
				ret = fread(&d.elapsedTime, sizeof(d.elapsedTime), 1, fd);
				ret = fread(&d.TotalExposureTime, sizeof(d.TotalExposureTime), 1, fd);
				if (verbose){
					printf("dataFileVersion = %d\n", d.dataFileVersion);
				}
				printf("numberOfScans = %d\n", d.numberOfScans);
				printf("elapsedTime = %.2lf\n", d.elapsedTime);
				printf("TotalExposureTime = %.2lf\n", d.TotalExposureTime);
				break;
			case DESCR_SECTION_ID :
				memcpy(&d.snapshotDescrHead, &header, sizeof(header));
				ret = fread(&d.sampleName, sizeof(d.sampleName), 1, fd);
				ret = fread(&d.sampleDescription, sizeof(d.sampleDescription), 1, fd);
				d.sampleName[99] = '\0';
				d.sampleDescription[1999] = '\0';
				printf("sample name: %s\n", d.sampleName);
				printf("sample Description:\n%s\n", d.sampleDescription);
				break;
			case REMOTE_DATA_SECTION_ID :
			default:
				break;
		}
		// Ici on jongle entre les bugs du fichier et ceux des structures
		// Si on n'a pas assez lu on fait confiance à la valeur de SectionLen du fichier
		// Si on a trop lu on fait confiance à la structure
		old_position = ftell(fd);
		if (old_position != position){
			if (verbose)
				fprintf(stderr, "WARNING: ecart de position old : %ld  %s new : %ld\n",
					old_position,
					old_position > position ? ">" : "<",
					position);
		}
		if (old_position > position){
			position = old_position;
		} else {
			fseek(fd, position, SEEK_SET);
		}
	}
	fclose(fd);
}

void main(int argc, char ** argv)
{
	int c;
	char * cvalue;
	int index;
#if 0
	printf("sizeof(int) = %ld\n", sizeof(int));
	printf("sizeof(long) = %ld\n", sizeof(long));
	printf("sizeof(float) = %ld\n", sizeof(float));
	printf("sizeof(double) = %ld\n", sizeof(double));
	printf("sizeof(SNAPSHOT_FILE_HEADER) = %ld\n", sizeof(SNAPSHOT_FILE_HEADER));
	printf("sizeof(SNAPSHOT_HEADER) = %ld\n", sizeof(SNAPSHOT_HEADER));
	printf("sizeof(struct tsDatFile) = %ld -- %08lo\n", sizeof(struct tsDatFile),  sizeof(struct tsDatFile));
#endif

   	double LeftHand[2][2][2][MAX_SAMPLES];
   	printf(" double LeftHand : %ld\n", sizeof(LeftHand));
   	printf(" double LeftHand[2] : %ld\n", sizeof(LeftHand[2]));
   	printf(" double LeftHand[2][2] : %ld\n", sizeof(LeftHand[2][2]));
   	printf(" double LeftHand[2][2][2] : %ld\n", sizeof(LeftHand[2][2][2]));
   	printf(" double LeftHand[2][2][2][MAX_SAMPLES] : %ld\n", sizeof(LeftHand[2][2][2][MAX_SAMPLES]));

	while((c = getopt(argc, argv, "hvz:+:")) != -1){
		switch(c){
			case 'v':
				verbose = 1;
				break;
			case 'h':
				//cvalue = optarg;
				break;
			case '+':
				opt_samples = atoi(optarg);
				break;
			case '?':
			default:
				if (optopt == 'z')
					fprintf(stderr,"Option -%c requires an argument.\n", optopt);
				else if (isprint(optopt))
					fprintf(stderr,"Unknown option '-%c'.\n", optopt);
				else
					fprintf(stderr,"Unknown character '\\x%x'.\n", optopt);
				return;

		}
	}
	for(index = optind; index < argc; index++){
		LoadSnapShot(argv[index]);
	}
}

File: roa.py
import struct
import sys
import numpy as np
from enum import IntEnum

class RoaConsts(IntEnum):
	MAX_SAMPLES             = 1024
	DATA_FILE_VERSION       = 1
	SECTION_SYNC_BYTE       = 0x40
	LOCAL_DATA_SECTION_ID   = 0x1111
	REMOTE_DATA_SECTION_ID  = 0x2222
	STATS_SECTION_ID        = 0x3333
	DESCR_SECTION_ID        = 0x4444
	FILE_SYNC_SEQ           = 0x414B53

class RoaFileHeader:
    def fromfile(self, fd):
      buffer = fd.read(struct.calcsize('<b3x'))
      self.VersionNum, = struct.unpack('<b3x', buffer)
      return self

class RoaHeader:
    def fromfile(self, fd):
      buffer = fd.read(struct.calcsize('<hbbi'))
      (self.SectionID,
        self.VersionNum,
        self.SyncByte,
        self.SectionLen) = struct.unpack('<hbbi', buffer)
      return self

#   SNAPSHOT_FILE_HEADER snapshotFileHead;
#   SNAPSHOT_HEADER snapshotDataHead;
#   float ROAspecta[MAX_SAMPLES];
#   float RAMANspecta[MAX_SAMPLES];
#   double LeftHand[2][2][2][MAX_SAMPLES];
#   double RightHand[2][2][2][MAX_SAMPLES];
#   float ROA_CC[MAX_SAMPLES];
#   float ROAnoiseFloor[MAX_SAMPLES];
#   SNAPSHOT_HEADER snapshotStatsHead;
#   int dataFileVersion;
#   int numberOfScans; 		// au depart long
#   double elapsedTime;		// au depart float
#   double TotalExposureTime;

def c2py_string(cstring):
  "Convertit une suite d'octets provenaire d'une chaine C en string python"
  s = ""
  for i in range(len(cstring)):
    if cstring[i] == 0:
      break
    s += chr(cstring[i])
  return s


class RoaData:
    def __init__(self):
      self.file_header = RoaFileHeader()
      self.file_header.VersionNum = 0
      self.RoaSpecta = ()
      self.RamanSpecta = ()
      self.LeftHand = ()
      self.RightHand = ()
      self.roa_cc = ()
      self.roaNoise = ()
      self.dataFileVersion = 0
      self.numberOfScans = 0
      self.elapsedTime = 0.0
      self.TotalExposureTime = 0.0
      self.name = ""
      self.description = ""

      self.fname = ""
      
    
    def open(self, fname):
      self.file_header.VersionNum = 0
      try:
        with open(fname, 'rb') as fd:
          self.fname = fname
          self.file_header = self.file_header.fromfile(fd)
  
          h = RoaHeader()
          h = h.fromfile(fd)
          if h.SectionID == int(RoaConsts.LOCAL_DATA_SECTION_ID):
            self.data_header = h
            # "Section de données, la longueur correcte de la section est dans h.SectionLen"
            buffer = fd.read(h.SectionLen)
            pos = 0
            self.RoaSpecta = struct.unpack_from('<' + str(int(RoaConsts.MAX_SAMPLES)) +'f', buffer, offset = pos)
            l = struct.calcsize('<' + str(int(RoaConsts.MAX_SAMPLES)) +'f')
            pos += l
            self.RamanSpecta = struct.unpack_from('<' + str(int(RoaConsts.MAX_SAMPLES)) +'f', buffer, offset = pos)
            l = struct.calcsize('<' + str(int(RoaConsts.MAX_SAMPLES)) +'f') 
            pos += l
            self.LeftHand = struct.unpack_from('<' +  str(2*2*2*int(RoaConsts.MAX_SAMPLES)) +'d', buffer, offset = pos)
            l = struct.calcsize('<' + str(2*2*2*int(RoaConsts.MAX_SAMPLES)) +'d') 
            pos += l
            self.RightHand = struct.unpack_from('<' + str(2*2*2*int(RoaConsts.MAX_SAMPLES)) +'d', buffer, offset = pos)
            l = struct.calcsize('<' + str(2*2*2*int(RoaConsts.MAX_SAMPLES)) +'d') 
            pos += l
            self.roa_cc = struct.unpack_from('<' + str(int(RoaConsts.MAX_SAMPLES)) +'f', buffer, offset = pos)
            l = struct.calcsize('<' + str(int(RoaConsts.MAX_SAMPLES)) +'f') 
            pos += l
            self.roaNoise = struct.unpack_from('<' + str(int(RoaConsts.MAX_SAMPLES)) +'f', buffer, offset = pos)
            l = struct.calcsize('<' + str(int(RoaConsts.MAX_SAMPLES)) +'f') 
            pos += l
  
          h = h.fromfile(fd)
          if h.SectionID == int(RoaConsts.STATS_SECTION_ID):
            self.stats_header = h
            "Section des statistiques, la longueur correcte de la section n'est pas dans h.SectionLen"
            "la section est plus grande que SectionLen"
            buffer = fd.read(struct.calcsize('<iidd'))
             
            (
              self.dataFileVersion,
              self.numberOfScans,
              self.elapsedTime,
              self.TotalExposureTime
            ) = struct.unpack('<iidd', buffer)
            
          h = h.fromfile(fd)
          if h.SectionID == int(RoaConsts.DESCR_SECTION_ID):
            self.descr_header = h
            buffer = fd.read(100)
            self.name = c2py_string(buffer)
            buffer = fd.read(2000)
            self.description = c2py_string(buffer)
  
        fd.close()
      except OSError as err:
        print("Error: {} loading file : {}\n".format(err, fname))
      
      except :
        print("Error : bad format in  file {}".format(fname))

      return self

    def getLeftHand(self,a, b, c, d):
      pos = ((a*2+b)*2+c) * int(RoaConsts.MAX_SAMPLES) + d
      if pos < len(self.LeftHand):
        return self.LeftHand[ pos]
      return -1.0  

    def getRightHand(self,a, b, c, d):
      pos = ((a*2+b)*2+c) * int(RoaConsts.MAX_SAMPLES) + d
      if pos < len(self.RightHand):
        return self.RightHand[ pos]
      return -1.0  
    
    def getRoaSpecta(self, pos):
      if pos < len(self.RoaSpecta):
        return self.RoaSpecta[pos]
      return -1.0

    def getRamanSpecta(self, pos):
      if pos < len(self.RamanSpecta):
        return int(self.RamanSpecta[pos])
      else:
        return -1
        
    def getRoaCC(self, pos):
      if pos < len(self.roa_cc):
        return self.roa_cc(pos)
      return -1.0

    def getRoaNoise(self, pos):
      if pos < len(self.roaNoise):
        return int(self.roaNoise)
      else:
        return -1
    
    def getNumberOfScans(self):
      return self.numberOfScans
    
    def getElapsedTime(self):
      return self.elapsedTime
      
    def getTotalExposureTime(self):
      return self.TotalExposureTime
    
    def getName(self):
      return self.name
    
    def getDescription(self):
      return self.description

    def getFileName(self):
      return self.fname
    
    def affiche(self, maximum = 0):
      if maximum > 0 :
            print('{:^4s};{:^15s};{:^15s}'.format('pos','raman', 'roa'))
      for i in range(min(maximum, int(RoaConsts.MAX_SAMPLES))):
        print('{:4d};{:15d};{:^11.4f}'.format(i,self.getRamanSpecta(i), self.getRoaSpecta(i)))

    def affiche_stats(self):
        print("{:^20s};{:6d};{:^8.4f};{:^8.4f}".format( self.name, self.numberOfScans, self.elapsedTime, self.TotalExposureTime))

import argparse
import os

def main(argv=sys.argv[1:]):
  parser = argparse.ArgumentParser(description='Process some ROA .dat files.')
  parser.add_argument('--showdata', default=False, action='store_true', help='Shown Content data')
  parser.add_argument('--showdesc', default=False, action='store_true', help='Shown Description')
  parser.add_argument('--showstats', default=False, action='store_true', help='Shown Statistics')
  parser.add_argument('-n', default=0, type=int, action='store', help='# elements to print')
  parser.add_argument('files', metavar='file', type=str, nargs='*',
                    help='show ROA content file')

  args = parser.parse_args()


  roa = RoaData()

  print(args)

  for fname in args.files:
    if not os.path.exists(fname):
        print("error: {} doesn't exist\n".format(fname))
    roa = roa.open(fname)
    # print(roa.getLeftHand(1,0,1,512))
    # print(roa.getName())
    if args.showdesc:
        print(roa.getDescription())
    if args.showstats:
        roa.affiche_stats()
    if args.showdata:
        roa.affiche(args.n)

if __name__ == '__main__':
    main()


